[CG] A large PDFDocumentImage may be displayed pixelated

shallawa opened the pull request:
#### 8382d4d3c6e3e691c2c1e1d0efc7d2a535898c93
<pre>
[CG] A large PDFDocumentImage may be displayed pixelated
<a href="https://bugs.webkit.org/show_bug.cgi?id=245664">https://bugs.webkit.org/show_bug.cgi?id=245664</a>
rdar://97254192

Reviewed by Simon Fraser.

Introduce a new class called CachedSubimage. This class will be responsible of
creating a local ImageBuffer, drawing the source rectangle of the Image to the
ImageBuffer then drawing this ImageBuffer to the destination GraphicsContext.

The geometry of CachedSubimage will be calculated such that the resolution of
the CachedSubimage ImageBuffer matches the effective resolution of the destination
retangle when it is drawm to GraphicsContext.

For GPU Process rendering, we have to draw the PDF image in WebProcess into a local
ImageBuffer, get a NativeImage from this ImageBuffer and then send it to GPUProcess.
CachedSubimage will be used to do this seemlessly for all PDFDocumentImages regardless
of their size.

If Image::shouldDrawFromCachedSubimage() is true, a CachedSubimage of the Image
with the current resolution will be generated. If the local ImageBuffer can&apos;t be
created and Image::mustDrawFromCachedSubimage() is true, a pixelated CachedSubimage
of the Image will be created. If the pixelated CachedSubimage could not be created,
the Image will not be drawn.

* LayoutTests/fast/images/cached-clipped-pdf-expected.html:
* LayoutTests/fast/images/cached-clipped-pdf.html:
* LayoutTests/fast/images/displaced-non-cached-pdf.html:
* LayoutTests/fast/images/pdf-as-image-dest-rect-change-expected.txt:
* LayoutTests/fast/images/pdf-as-image-dest-rect-change.html:
* LayoutTests/fast/images/pdf-in-canvas.html:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/page/PDFImageCachingPolicy.h: Removed.
* Source/WebCore/page/Settings.yaml:
* Source/WebCore/page/SettingsBase.h:
* Source/WebCore/platform/graphics/CachedSubimage.cpp: Added.
(WebCore::CachedSubimage::create):
(WebCore::calculateCachedSubimageSourceRect):
(WebCore::CachedSubimage::createCachedSubimage):
(WebCore::CachedSubimage::createPixelatedCachedSubimage):
(WebCore::CachedSubimage::CachedSubimage):
(WebCore::CachedSubimage::canBeUsed const):
(WebCore::CachedSubimage::draw):
* Source/WebCore/platform/graphics/CachedSubimage.h: Added.
* Source/WebCore/platform/graphics/FloatRect.h:
(WebCore::areEssentiallyEqual):
* Source/WebCore/platform/graphics/GeneratedImage.h:
* Source/WebCore/platform/graphics/GraphicsContext.cpp:
(WebCore::GraphicsContext::drawImage):
* Source/WebCore/platform/graphics/Image.cpp:
(WebCore::Image::drawCachedSubimage):
(WebCore::Image::destroyDecodedData):
* Source/WebCore/platform/graphics/Image.h:
(WebCore::Image::cachedSubimageCreateCountForTesting const):
(WebCore::Image::cachedSubimageDrawCountForTesting const):
(WebCore::Image::shouldDrawFromCachedSubimage const):
(WebCore::Image::mustDrawFromCachedSubimage const):
(WebCore::Image::drawCachedSubimage):
* Source/WebCore/platform/graphics/cg/PDFDocumentImage.cpp:
(WebCore::PDFDocumentImage::decodedSizeChanged):
(WebCore::PDFDocumentImage::shouldDrawFromCachedSubimage const):
(WebCore::PDFDocumentImage::mustDrawFromCachedSubimage const):
(WebCore::PDFDocumentImage::draw):
(WebCore::PDFDocumentImage::destroyDecodedData):
(WebCore::PDFDocumentImage::createPDFDocument):
(WebCore::PDFDocumentImage::setPdfImageCachingPolicy): Deleted.
(WebCore::PDFDocumentImage::cacheParametersMatch const): Deleted.
(WebCore::transformContextForPainting): Deleted.
(WebCore::cachedImageRect): Deleted.
(WebCore::PDFDocumentImage::updateCachedImageIfNeeded): Deleted.
* Source/WebCore/platform/graphics/cg/PDFDocumentImage.h:
* Source/WebCore/platform/mediastream/WebRTCProvider.h:
* Source/WebCore/rendering/RenderImage.cpp:
(WebCore::RenderImage::paintIntoRect):
* Source/WebCore/svg/graphics/SVGImage.h:
* Source/WebCore/svg/graphics/SVGImageForContainer.h:
* Source/WebCore/testing/InternalSettings.cpp:
(WebCore::InternalSettings::Backup::Backup):
(WebCore::InternalSettings::Backup::restoreTo):
(WebCore::InternalSettings::setStorageBlockingPolicy):
(WebCore::InternalSettings::setPDFImageCachingPolicy): Deleted.
* Source/WebCore/testing/InternalSettings.h:
* Source/WebCore/testing/InternalSettings.idl:
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::imageCachedSubimageCreateCount):
(WebCore::pdfDocumentImageFromImageElement): Deleted.
(WebCore::Internals::pdfDocumentCachingCount): Deleted.
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:

Canonical link: <a href="https://commits.webkit.org/259928@main">https://commits.webkit.org/259928@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bc6d95bc0bcbaa8f787b667efa93baea36bb7b44

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/106003 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/15058 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/38835 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/115186 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/175294 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/109907 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/16482 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/6242 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/98212 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/114923 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/111756 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/12547 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/95517 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/40089 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/94412 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/27160 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/81758 "Found 1 new API test failure: /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/text/state-changed (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/8314 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/28512 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/8804 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/5076 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/14428 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/48055 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6860 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/10348 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/platform/graphics/ImageViewport.h:
```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright (C) 2022 Apple Inc.  All rights reserved.
```
2023

shallawa replied:
Fixed.

smfr commented on Source/WebCore/platform/graphics/cg/PDFDocumentImage.cpp:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (C) 2004, 2005, 2006, 2013 Apple Inc.  All rights reserved.
+ * Copyright (C) 2004-2022 Apple Inc.  All rights reserved.
```
2023

shallawa replied:
Fixed.

smfr commented on Source/WebCore/platform/graphics/cg/PDFDocumentImage.h:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (C) 2004, 2005, 2006, 2013 Apple Inc.  All rights reserved.
+ * Copyright (C) 2004-2022 Apple Inc.  All rights reserved.
```
2023

shallawa replied:
Fixed.

smfr commented on Source/WebCore/testing/Internals.idl:
```diff
@@ -567,7 +567,7 @@ typedef (FetchRequest or FetchResponse) FetchObject;
     undefined setClearDecoderAfterAsyncFrameRequestForTesting(HTMLImageElement element, boolean enabled);
     unsigned long imageDecodeCount(HTMLImageElement element);
-    unsigned long pdfDocumentCachingCount(HTMLImageElement element);
+    unsigned long viewportCreateCount(HTMLImageElement element);
```
this needs "imageViewport" in the name, because "viewport" can mean lots of things (visual viewport, layout viewport etc).

shallawa replied:
Fixed.

smfr commented on Source/WebCore/platform/graphics/Image.h:
```diff
@@ -203,7 +208,10 @@ class Image : public RefCounted<Image> {
 #if PLATFORM(WIN)
     virtual void drawFrameMatchingSourceSize(GraphicsContext&, const FloatRect& dstRect, const IntSize& srcSize, CompositeOperator) { }
 #endif
+    virtual bool canDrawViewport(GraphicsContext&) const { return false; }
```
I think using "viewport" here is confusing (it assumes knowledge of some internal behavior of Image). It's more about "canUseCachedRepresentation()" or something.

shallawa replied:
I will think about a better name.

smfr commented on Source/WebCore/platform/graphics/Image.h:
```diff
@@ -203,7 +208,10 @@ class Image : public RefCounted<Image> {
     virtual void drawFrameMatchingSourceSize(GraphicsContext&, const FloatRect& dstRect, const IntSize& srcSize, CompositeOperator) { }
 #endif
+    virtual bool canDrawViewport(GraphicsContext&) const { return false; }
+    virtual bool shouldDrawViewport(GraphicsContext&) const { return false; }
```
Same

smfr commented on Source/WebCore/platform/graphics/Image.h:
```diff
@@ -203,7 +208,10 @@ class Image : public RefCounted<Image> {
+    virtual bool canDrawViewport(GraphicsContext&) const { return false; }
+    virtual bool shouldDrawViewport(GraphicsContext&) const { return false; }
     virtual ImageDrawResult draw(GraphicsContext&, const FloatRect& dstRect, const FloatRect& srcRect, const ImagePaintingOptions& = { }) = 0;
+    ImageDrawResult drawViewport(GraphicsContext&, const FloatRect& dstRect, const FloatRect& srcRect, const ImagePaintingOptions& = { });
```
Can this be private and ImageViewport made a friend class? Having this public makes things confusing.

shallawa replied:
`drawViewport()` is a protected method of `Image`. And `ImageViewport` is a friend of `Image`.

smfr commented on Source/WebCore/platform/graphics/Image.h:
```diff
@@ -213,10 +221,14 @@ class Image : public RefCounted<Image> {
     std::optional<bool> m_allowsAnimation { std::nullopt };
+    std::unique_ptr<Timer> m_animationStartTimer;
+
+    std::unique_ptr<ImageViewport> m_viewport;
```
m_imageViewport.

shallawa replied:
Fixed.

smfr commented on Source/WebCore/platform/graphics/ImageViewport.h:
```diff
@@ -0,0 +1,59 @@
+class ImageBuffer;
+struct ImagePaintingOptions;
+
+class ImageViewport {
```
I'm not a fan of the "viewport" name since we use that term in so many other places. This is more like a cached subimage.

smfr commented on Source/WebCore/platform/graphics/cg/PDFDocumentImage.cpp:
```diff
@@ -97,94 +80,6 @@ EncodedDataStatus PDFDocumentImage::dataChanged(bool allDataReceived)
-}
-
-// To avoid the jetsam on iOS, we are going to limit the size of all the PDF cachedImages to be 64MB.
-static const size_t s_maxCachedImageSide = 4 * 1024;
```
Do we just not need this code any more? Are we going to trigger more iOS jetsams by removing it?

shallawa replied:
I do not think so. The maximum area of ImageViewport in device pixels is 
```
max_viewport_area = ImageBuffer::maxClampedSide x ImageBuffer::maxClampedSide
```
which is the same as the old code see `calculateViewportSourceRect()`.

Also this maximum area 64MB is the maximum of the 2D canvas on iOS, see `maxCanvasArea()`. So if we are allowing this size for a 2D canvas on iOS, the PDF image will have the same memory cost. But this is in addition to the encoded data of the PDF file and the native image we create for the image which is drawn on the tile.

shallawa replied:
I made the maximum size of the ImageBuffer stored by CachedSubimage in bytes:

```
    // The tile size is usually 512 x 512, account for the 2x display resolution
    // and make room for one more tile in every direction.
    static constexpr float maxSide = 512 * 2 * 3
    static constexpr float maxArea = maxSide * maxSide;
    static constexpr float maxMemory = maxArea * 4;
```

This is 36MB. And this gives a higher cache usage which is a little bit more than 2. (create 100 sublimates and draw 200 times).

smfr commented on Source/WebCore/platform/graphics/CachedSubimage.h:
```diff
@@ -0,0 +1,64 @@
+
+    // The tile size is usually 512 x 512, account for the 2x display resolution
+    // and make room for one more tile in every direction.
+    static constexpr float maxSide = 512 * 2 * 3;
```
Should this be more dynamic? We have 3x displays on iPhones.

shallawa replied:
Yes I think we should. 

But we have a problem getting the `deviceScaleFactor` from the `GraphicsContext`. `GraphicsContext::scaleFactor()` returns the user transform multiplied by the `deviceScaleFactor`. Also `GraphicsContextCG::getCTM()` takes the argument `IncludeDeviceScale` whose values are confusing 
```
 enum IncludeDeviceScale { DefinitelyIncludeDeviceScale, PossiblyIncludeDeviceScale }
```
But even with passing PossiblyIncludeDeviceScale on WK1, I am still getting the user transform multiplied by the `deviceScaleFactor`. 

And this is why GraphicsContextCG uses the local inline function `getUserToBaseCTM()` to get the user CTM without the `deviceScaleFactor`. Also ControlMac::drawCell() has a duplicate of the same calculation to know whether to use an ImageBuffer or not.

Because this needs cleaning up the `GraphicsContext::scaleFactor()` and `GraphicsContext::getCTM()` and removing the duplicate code in GraphicsContextCG.cpp and ControlMac::drawCell(), I am going to address this in a separate patch.

shallawa committed:
[CG] A large PDFDocumentImage may be displayed pixelated
https://bugs.webkit.org/show_bug.cgi?id=245664
rdar://97254192

Reviewed by Simon Fraser.

Introduce a new class called CachedSubimage. This class will be responsible of
creating a local ImageBuffer, drawing the source rectangle of the Image to the
ImageBuffer then drawing this ImageBuffer to the destination GraphicsContext.

The geometry of CachedSubimage will be calculated such that the resolution of
the CachedSubimage ImageBuffer matches the effective resolution of the destination
retangle when it is drawm to GraphicsContext.

For GPU Process rendering, we have to draw the PDF image in WebProcess into a local
ImageBuffer, get a NativeImage from this ImageBuffer and then send it to GPUProcess.
CachedSubimage will be used to do this seemlessly for all PDFDocumentImages regardless
of their size.

If Image::shouldDrawFromCachedSubimage() is true, a CachedSubimage of the Image
with the current resolution will be generated. If the local ImageBuffer can't be
created and Image::mustDrawFromCachedSubimage() is true, a pixelated CachedSubimage
of the Image will be created. If the pixelated CachedSubimage could not be created,
the Image will not be drawn.

* LayoutTests/fast/images/cached-clipped-pdf-expected.html:
* LayoutTests/fast/images/cached-clipped-pdf.html:
* LayoutTests/fast/images/displaced-non-cached-pdf.html:
* LayoutTests/fast/images/pdf-as-image-dest-rect-change-expected.txt:
* LayoutTests/fast/images/pdf-as-image-dest-rect-change.html:
* LayoutTests/fast/images/pdf-in-canvas.html:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/page/PDFImageCachingPolicy.h: Removed.
* Source/WebCore/page/Settings.yaml:
* Source/WebCore/page/SettingsBase.h:
* Source/WebCore/platform/graphics/CachedSubimage.cpp: Added.
(WebCore::CachedSubimage::create):
(WebCore::calculateCachedSubimageSourceRect):
(WebCore::CachedSubimage::createCachedSubimage):
(WebCore::CachedSubimage::createPixelatedCachedSubimage):
(WebCore::CachedSubimage::CachedSubimage):
(WebCore::CachedSubimage::canBeUsed const):
(WebCore::CachedSubimage::draw):
* Source/WebCore/platform/graphics/CachedSubimage.h: Added.
* Source/WebCore/platform/graphics/FloatRect.h:
(WebCore::areEssentiallyEqual):
* Source/WebCore/platform/graphics/GeneratedImage.h:
* Source/WebCore/platform/graphics/GraphicsContext.cpp:
(WebCore::GraphicsContext::drawImage):
* Source/WebCore/platform/graphics/Image.cpp:
(WebCore::Image::drawCachedSubimage):
(WebCore::Image::destroyDecodedData):
* Source/WebCore/platform/graphics/Image.h:
(WebCore::Image::cachedSubimageCreateCountForTesting const):
(WebCore::Image::cachedSubimageDrawCountForTesting const):
(WebCore::Image::shouldDrawFromCachedSubimage const):
(WebCore::Image::mustDrawFromCachedSubimage const):
(WebCore::Image::drawCachedSubimage):
* Source/WebCore/platform/graphics/cg/PDFDocumentImage.cpp:
(WebCore::PDFDocumentImage::decodedSizeChanged):
(WebCore::PDFDocumentImage::shouldDrawFromCachedSubimage const):
(WebCore::PDFDocumentImage::mustDrawFromCachedSubimage const):
(WebCore::PDFDocumentImage::draw):
(WebCore::PDFDocumentImage::destroyDecodedData):
(WebCore::PDFDocumentImage::createPDFDocument):
(WebCore::PDFDocumentImage::setPdfImageCachingPolicy): Deleted.
(WebCore::PDFDocumentImage::cacheParametersMatch const): Deleted.
(WebCore::transformContextForPainting): Deleted.
(WebCore::cachedImageRect): Deleted.
(WebCore::PDFDocumentImage::updateCachedImageIfNeeded): Deleted.
* Source/WebCore/platform/graphics/cg/PDFDocumentImage.h:
* Source/WebCore/platform/mediastream/WebRTCProvider.h:
* Source/WebCore/rendering/RenderImage.cpp:
(WebCore::RenderImage::paintIntoRect):
* Source/WebCore/svg/graphics/SVGImage.h:
* Source/WebCore/svg/graphics/SVGImageForContainer.h:
* Source/WebCore/testing/InternalSettings.cpp:
(WebCore::InternalSettings::Backup::Backup):
(WebCore::InternalSettings::Backup::restoreTo):
(WebCore::InternalSettings::setStorageBlockingPolicy):
(WebCore::InternalSettings::setPDFImageCachingPolicy): Deleted.
* Source/WebCore/testing/InternalSettings.h:
* Source/WebCore/testing/InternalSettings.idl:
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::imageCachedSubimageCreateCount):
(WebCore::pdfDocumentImageFromImageElement): Deleted.
(WebCore::Internals::pdfDocumentCachingCount): Deleted.
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:

Canonical link: https://commits.webkit.org/259928@main
